components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaAccountManager.kt

```diff
@@ -0,0 +1,423 @@
+     * Account was successfully authenticated.
+     * @param account An authenticated instance of a [FirefoxAccountShaped].
+     */
+    fun onAuthenticated(account: FirefoxAccountShaped)
+
+    /**
+     * Account's profile is now available.
+     * @param profile A fresh version of account's [Profile].
+     */
+    fun onProfileUpdated(profile: Profile)
+
+    /**
+     * Account manager encountered an error. Inspect [error] for details.
+     * @param error A specific FxA error encountered.
+     */
+    fun onError(error: FxaException)
+}
+
+/**
+ * An account manager which encapsulates various internal details of an account lifecycle and provides
+ * an observer interface along with a public API for interacting with an account.
+ * The internal state machine abstracts over state space as exposed by the fxaclient library, not
+ * the internal states experienced by lower-level representation of a Firefox Account; those are opaque to us.
+ *
+ * Class is 'open' to facilitate testing.
+ *
+ * @param context A [Context] instance that's used for internal messaging and interacting with local storage.
+ * @param config A [Config] used for account initialization.
+ * @param scopes A list of scopes which will be requested during account authentication.
+ */
+open class FxaAccountManager(
+    private val context: Context,
+    private val config: Config,
+    private val scopes: Array<String>,
+    private val accountStorage: AccountStorage = SharedPrefAccountStorage(context)
+) : Closeable, Observable<AccountObserver> by ObserverRegistry() {
+    private val logTag = "FirefoxAccountStateMachine"
+
+    private val oauth = object : Observable<OAuthObserver> by ObserverRegistry() {}
+
+    companion object {
+        /**
+         * State transition matrix. It's in the companion object to enforce purity.
+         * @return An optional [AccountState] if provided state+event combination results in a
+         * state transition. Note that states may transition into themselves.
+         */
+        @Suppress("ComplexMethod")
+        fun nextState(state: AccountState, event: AccountEvent): AccountState? {
+            return when (state) {
+                AccountState.Start -> {
+                    when (event) {
+                        Init -> AccountState.Start
+                        AccountNotFound -> AccountState.NotAuthenticated
+                        AccountRestored -> AccountState.AuthenticatedNoProfile
+                        else -> null
+                    }
+                }
+                AccountState.NotAuthenticated -> {
+                    when (event) {
+                        Authenticate -> AccountState.NotAuthenticated
+                        FailedToAuthenticate -> AccountState.NotAuthenticated
+                        is Authenticated -> AccountState.AuthenticatedNoProfile
+                        else -> null
+                    }
+                }
+                AccountState.AuthenticatedNoProfile -> {
```

Comment:
Is there any way we can do to get rid of this state? I think we were hoping to treat the failure to get profile as a non-logged in case. 
cc @eoger

Reply:
You could fail here for many reasons not related to authentication (intermittent networking problems for the device, server, profile endpoint issues, whatever else). At the same time, Sync could continue to work happily.
A better approach would be to capture explicitly auth-related problems while fetching a profile as events and have the state machine deal with them.

Reply:
I tend to agree with @vladikoff, I think having a profile shouldn't be a state: in a way profile fetching is its own little state machine (no profile, loading, has profile, error fetch etc).
I'm also concerned about the upcoming combinations of states in the future (Is logged in, has profile, has push subscription but needs to register its instance again).